java/arcs/core/data/expression/PaxelParser.kt

```diff
@@ -0,0 +1,292 @@
+import arcs.core.util.div
+import arcs.core.util.eof
+import arcs.core.util.many
+import arcs.core.util.map
+import arcs.core.util.optional
+import arcs.core.util.parser
+import arcs.core.util.plus
+import arcs.core.util.regex
+import arcs.core.util.token
+import arcs.core.util.unaryMinus
+import java.math.BigInteger
+
+/**
+ * A parser combinator implementation of the Paxel expression parser.
+ */
+object PaxelParser {
+    private val WS = "[\\s\\n]"
+
+    sealed class DiscreteType<T : Number> {
+        object PaxelBigInt : DiscreteType<BigInteger>() {
+            override fun parse(number: String) = BigInteger(number)
+        }
+
+        object PaxelInt : DiscreteType<Int>() {
+            override fun parse(number: String) = number.toInt()
+        }
+
+        object PaxelLong : DiscreteType<Long>() {
+            override fun parse(number: String) = number.toLong()
+        }
+
+        abstract fun parse(number: String): T
+    }
+
+    sealed class Unit(val conversionFactor: Int) {
+        fun convert(number: Number): Number = when (number) {
+            is BigInteger -> number.multiply(conversionFactor.toBigInteger())
+            is Long -> number * conversionFactor
+            is Double -> number * conversionFactor.toDouble()
+            else -> number.toLong() * conversionFactor
+        }
+
+        object Millisecond : Unit(1)
+        object Second : Unit(1000)
+        object Minute : Unit(Second.conversionFactor * 60)
+        object Hour : Unit(Minute.conversionFactor * 60)
+        object Day : Unit(Hour.conversionFactor * 24)
+    }
+
+    private val whitespace = -regex("($WS+)")
+
+    private val units =
+        optional(whitespace + (regex("(millisecond)s?").map { Unit.Millisecond } /
+        regex("(second)s?").map { Unit.Second } /
+        regex("(minute)s?").map { Unit.Minute } /
+        regex("(hour)s?").map { Unit.Hour } /
+        regex("(day)s?").map { Unit.Day })).map {
+            it ?: Unit.Millisecond
+        }
+
+    private val typeIdentifier =
+        token("n").map { DiscreteType.PaxelBigInt } /
+        token("i").map { DiscreteType.PaxelInt } /
+        token("l").map { DiscreteType.PaxelLong }
+
+    private val discreteValue =
+        (regex("(-?[0-9]+)") + typeIdentifier + units).map { (number, type, units) ->
+            Expression.NumberLiteralExpression(units.convert(type.parse(number)))
+        }
+
+    private val numberValue = (regex("([+-]?[0-9]+\\.?[0-9]*(?:[eE][+-]?[0-9]+)?)") + units)
+        .map { (number, units) ->
+            Expression.NumberLiteralExpression(units.convert(number.toDouble()))
+        }
+
+    private val booleanValue =
+        token("true").map { true.asExpr() } / token("false").map { false.asExpr() }
+
+    private val textValue = regex("\'((?:[^\'\\\\]|\\\\[\'\\\\/bfnrt]|\\\\u[0-9a-f]{4})*)\'").map {
+        Expression.TextLiteralExpression(unescape(it))
+    }
+
+    private val ident = regex("([A-Za-z_][A-Za-z0-9_]*)")
```

Comment:
Looks like a lot of the 'optional whitespace' (ows) could be folded into the identifier parser. Might save some code size +/ avoid some bugs where we forget to put in an `ows`?